Content.Server/Chat/Systems/ChatSystem.Events.cs

```diff
@@ -0,0 +1,211 @@
+        /// </summary>
+        /// <remarks>
+        /// The objects are passed to the recipients at a later stage of processing.
+        /// </remarks>
+        public Dictionary<EntityUid, EntityChatData> Recipients = new();
+
+        /// <summary>
+        /// The designated ChatChannel for this message.
+        /// </summary>
+        public ChatChannel Channel;
+
+        /// <summary>
+        /// Which system, if any, has claimed this chat message.
+        /// </summary>
+        /// <remarks>
+        /// If a system claims a chat, they are saying they will be responsible for sending it.
+        /// </remarks>
+        public Type? ClaimedBy;
+
+        public EntityChat(EntityUid source, string message)
+        {
+            Source = source;
+            Message = message;
+        }
+    }
+
+    // New Chat flow:
+    //
+    // Parse                    - we have no idea what we have
+    // Attempt                  - now we know what type of chat this is, do we continue?
+    // GetRecipients            - we get who will receive this chat
+    // Transform                - we can change what the speaker says
+    //
+    // BeforeEntityChat         - we could cancel the chat at the recipient
+    // GotEntityChatTransform   - we can change what the recipient hears
+    // GotEntityChat            - finally, show the user the message
+
+    [ByRefEvent]
+    public struct EntityChatParseEvent
+    {
+        public EntityChat Chat;
+
+        public EntityChatParseEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public struct EntityChatAttemptEvent
+    {
+        public EntityChat Chat;
+
+        public EntityChatAttemptEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Cancelled { get; private set; }
+        public void Cancel() => Cancelled = true;
+        public void Uncancel() => Cancelled = false;
+    }
+
+    [ByRefEvent]
+    public struct EntityChatGetRecipientsEvent
+    {
+        public readonly EntityChat Chat;
+
+        public EntityChatGetRecipientsEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public struct EntityChatTransformEvent
+    {
+        public EntityChat Chat;
+
+        public EntityChatTransformEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public struct BeforeEntityChatEvent
+    {
+        public readonly EntityUid Recipient;
+        public readonly EntityChat Chat;
+
+        public BeforeEntityChatEvent(EntityUid recipient, EntityChat chat)
+        {
+            Recipient = recipient;
+            Chat = chat;
+        }
+
+        public bool Cancelled { get; private set; }
+        public void Cancel() => Cancelled = true;
+        public void Uncancel() => Cancelled = false;
```

Comment:
This is a really weird use of encapsulation that will never be used again, just let the Cancelled bool be set directly.